### PR TITLE
Threejs demo

### PR DESCRIPTION
**Pull request by patrick-volograms**
## Purpose of this PR

Replace this text with a short, clear, description of **_why_** you did this work in plain English.

This PR adds a three.js example to the project. The demo is based off the wasm webgl example, but the gl functions are replaces with three.js equivalents.

## Changes made in this PR

Write a summary here describing **_what_** work you did in your branch, and **_how_** you did it (explain your approach).

* Created a new project in the examples folder: 05_vol_geom_wasm_threejs
* The example 04 project was copied into the new example, with web gl functions remove
* Added the three.js file, but also put in a comment of how to import it with CDN
* Added a function to retrieve the max blob size of a vologram in wasm_vol_geom.c (used to initialise the mesh buffers in three js)
* The layout of the new example was made to match example 04

## Testing Summary

Provide more detail here as required.

- [x] Mac with Chrome - Works fine
- [ ] Mac with Safari - `TypeError: texture_video_el.requestVideoFrameCallback is not a function.` (Not good, need to check)
- [x] iPad Air with Chrome - Works mostly fine, some texture desync (seems to consistently desync around same time, see video) 
- [x] iPad Air with Safari - Same as iPad Air Chrome (not surprising, similar backends) 

## Risk Introduced and any Breaking Changes

Provide more detail here as required.

- [x] None as far as I can tell

## Pre-Merge Checklist

* If you have a problem with the review process (e.g. it's taking too long, or the reviewers forgot) get in touch with the Team Lead of the project!

- [ ] Before making your PR, refresh your branch with a merge from its parent in case of conflicting changes since you started your branch.
- [ ] This PR description is written clearly so that reviewers can understand the why/what/how of the work done.
- [ ] Reviewed your own PR and checked your changed code.
- [ ] Invited a range of good reviewers.
- [ ] Wait for a PR to be properly reviewed. Responded to all questions, and made any required changes.
- [ ] Received one other reviewer's approval. Ideally wait for 2+ approvals, especially if you didn't get a thorough reivew on the first approval.
- [ ] If a reviewer asks a question, even if you have another approval, got the approval of the other reviewers as well before merging.
- [ ] Made sure existing reviewers approved any changes to the PR.
- [ ] Resolved any merge conflicts after making the PR.
- [ ] CI/CD builds are not reporting any errors.

[Video of iPad Chrome 
](https://youtu.be/YVJ3kJf7AUo)

### CODE REVIEW

File: examples/05_vol_geom_wasm_threejs/index.html [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+	vologram.video_url = "./calif/output_720.mp4";
+	vologram.fps = 30.0;           // Change this value if not using 30 frames-per-second video.
+	var init_done = false;         // If the basic shaders/objects are set up. A vologram can be downloaded & prepared after this is true.
+
+	function create_vologram_geometry() {
+		vologram.geometry = new THREE.BufferGeometry();
+		vologram.geometry.setAttribute(
+			"position", 
+			new THREE.Float32BufferAttribute(
+				new Float32Array(vologram.max_blob_size), 3, false
```
Comment (capnramses) [inline examples/05_vol_geom_wasm_threejs/index.html:56]: i tested it after removing these buffer pre-allocations and it worked fine. i'm not sure if it makes it faster/slower - i didnt see any desync when i tested it.

File: examples/05_vol_geom_wasm_threejs/index.html [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+	function create_vologram_mesh() {
+		vologram.material = new THREE.MeshBasicMaterial({
+			color: 0xffffff,
+			map: vologram.texture,
+		});
+
+		vologram.mesh = new THREE.Mesh(vologram.geometry, vologram.material);
+		vologram.mesh.position.set( 0, 0, 0 )
+		// The vologram is mirrored
+		vologram.mesh.scale.set( -1, 1, 1 )
```
Comment (capnramses) [inline examples/05_vol_geom_wasm_threejs/index.html:92]: might need to do this in the mesh update function rather, since it seems to get wiped out after each update.

the rotation only wants to be applied once though so goes here.

File: examples/05_vol_geom_wasm_threejs/index.html [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+
+			// Indices - fetch and upload.
+			var indices_copied = Module.ccall('frame_indices_copied', 'number');
+			var indices_sz = Module.ccall('frame_i_sz', 'number');
+			vologram.n_indices = indices_sz / 2; // ushort
+			var indices_u16 = new Uint16Array(Module.HEAP8.buffer, indices_copied, vologram.n_indices);
+			vologram.geometry.setIndex( new THREE.Uint16BufferAttribute( indices_u16, 1 ))
+		}
+
+		// It seems that calculating bounding sphere does not work and always returns NaN value for the radius
```
Comment (capnramses) [inline examples/05_vol_geom_wasm_threejs/index.html:214]: i bet the NaN is when there is nobody in the scene. could probably check the vertex count > 0 before doing the bounding sphere update.

File: examples/05_vol_geom_wasm_threejs/index.html [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+		console.log( "max blob size =", vologram.max_blob_size );
+		create_vologram();
+		console.log("vologram.has_normals = " + vologram.has_normals);
+		mesh_from_frame(0); // Load in frame 0 so something displays even if we don't play yet.
+		console.log("mesh initialised");
+		start_video();
+		vologram.ready = true;
+	}
+
+	// This can be called to add an artificial slow-down in the code. Parameter value is in micro-seconds.
```
Comment (capnramses) [inline examples/05_vol_geom_wasm_threejs/index.html:256]: if we call this function in different places when testing we can test the robustness vs desync.

File: examples/05_vol_geom_wasm_threejs/index.html [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+		vologram.geometry.setIndex(
+			new THREE.Uint16BufferAttribute(
+				new Uint16Array(vologram.max_blob_size), 1, false
+			)
+		);
+	}
+
+	function create_vologram_texture() {
+		var texture_video_el = document.getElementById("texture_video_el");
+		vologram.texture = new THREE.VideoTexture( texture_video_el );
```
Comment (capnramses) [inline examples/05_vol_geom_wasm_threejs/index.html:80]: if this is introducing some desync we could do a manual update of the texture inside the video callback. we can keep testing this example after merge to figure that out

File: examples/05_vol_geom_wasm_threejs/index.html [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+		}
+
+		console.log("init done");
+		init_done = true;
+
+		return true;
+	}
+
+	function render() {
+		renderer.render(scene, camera);
```
Comment (capnramses) [inline examples/05_vol_geom_wasm_threejs/index.html:154]: one caveat with mirroring the vologram is that we need to switch webgl to render in clockwise winding order. i couldnt see anywhere to do that in three.js on a per-geometry or per-material basis, so last time i just set it globally in webgl. if that's required we probably want to put the vologram in its own scene, and any other geometry in a normal CCW scene and render them one after the other.

**Review by capnramses — DISMISSED** [review]
looks great!
just put some thoughts inline but i couldnt spot any bugs (except a guess at why the NaN thing happens).

**Comment by capnramses** [comment]
oh, we should add this to the table of examples on the ReadMe.md...

File: examples/05_vol_geom_wasm_threejs/index.html (hunk 1 above)
Comment (patrick-volograms) [reply examples/05_vol_geom_wasm_threejs/index.html:56]: Removing it, the less code the better 😅

File: examples/05_vol_geom_wasm_threejs/index.html (hunk 2 above)
Comment (patrick-volograms) [reply examples/05_vol_geom_wasm_threejs/index.html:92]: That seems strange, I'll try setting the scale of the geometry scale rather than the mesh and see if that's permanent

File: examples/05_vol_geom_wasm_threejs/index.html (hunk 3 above)
Comment (patrick-volograms) [reply examples/05_vol_geom_wasm_threejs/index.html:214]: For this example, we don't need the bounding sphere, but I don't know if its NaN because of a vologram issue, or a threejs issue

File: examples/05_vol_geom_wasm_threejs/index.html (hunk 4 above)
Comment (patrick-volograms) [reply examples/05_vol_geom_wasm_threejs/index.html:256]: I left it in for that. Could be handy in the future for unit/fuzz testing

File: examples/05_vol_geom_wasm_threejs/index.html (hunk 5 above)
Comment (patrick-volograms) [reply examples/05_vol_geom_wasm_threejs/index.html:80]: https://threejs.org/docs/#manual/en/introduction/How-to-update-things

I think that's possible

**Review by capnramses — APPROVED** [review]
